### PR TITLE
[opentelemetry-demo] Adds schedulingRules with nodeSelector, nodeAffinity and tolerations 

### PR DESCRIPTION
**Pull request by ugosan**
When using this chart on tainted nodes (e.g. with ARM `kubernetes.io/arch=arm64:NoSchedule`), we must add a nodeSelector to the deployments, this PR adds the ability to specify a `nodeSelector` like so:

```yml
  paymentService:
   ...
    podAnnotations: {}    
    nodeSelector:
      kubernetes.io/arch: arm64
```

### CODE REVIEW

File: charts/opentelemetry-demo/values.schema.json [hunk 1]
```diff
@@ -125,6 +125,9 @@
         },
         "podAnnotations": {
           "type": "object"
+        },
+        "nodeSelector": {
```
Comment (TylerHelmuth) [inline charts/opentelemetry-demo/values.schema.json:129]: Please add an empty `nodeSelector` property in the values.yaml, like `podAnnotations`, that helps users understand when this field would be used.

**Review by TylerHelmuth — COMMENTED** [review]
@ugosan thanks for the contribution.  Please bump the patch version of the chart.

File: charts/opentelemetry-demo/values.schema.json (hunk 1 above)
Comment (ugosan) [reply charts/opentelemetry-demo/values.schema.json:129]: makes sense, all done!

**Review by Allex1 — APPROVED** [review]

**Review by TylerHelmuth — COMMENTED** [review]
@ugosan were you able to test locally to confirm these changes work as expected?

**Comment by dmitryax** [comment]
If we are adding nodeSelector, I believe we also need `affinity` and `tolerations`.

**Comment by ugosan** [comment]
@TylerHelmuth Yes, I have the opentelemetry demo deployed on a cluster with two node pools (arm and amd)
```
NAME                                            STATUS    NODE-SELECTOR
aiops-ad-service-6c89dbfb95-9nsms               Running   map[kubernetes.io/arch:arm64]
aiops-cart-service-57b6cf696b-l2rg2             Running   map[kubernetes.io/arch:amd64]
aiops-checkout-service-b77bccdd9-6hptk          Running   map[kubernetes.io/arch:arm64]
aiops-currency-service-55b975fd7f-j9vj6         Running   map[kubernetes.io/arch:arm64]
aiops-email-service-56657f7c5b-dcgq4            Running   map[kubernetes.io/arch:arm64]
aiops-featureflag-service-668fbdd6f4-s6584      Running   map[kubernetes.io/arch:arm64]
aiops-ffs-postgres-6f5976774-gx2wn              Running   map[kubernetes.io/arch:arm64]
aiops-frontend-66988d6f9-hr2z8                  Running   map[kubernetes.io/arch:arm64]
aiops-jaeger-f7df8fbf9-hvlwd                    Running   <none>
aiops-loadgenerator-5f5659f859-7b25t            Running   map[kubernetes.io/arch:arm64]
aiops-otelcol-75c5f696b8-855vp                  Running   <none>
aiops-payment-service-78674fc8f5-frlg5          Running   map[kubernetes.io/arch:arm64]
aiops-product-catalog-service-998bdfc7f-dls2q   Running   map[kubernetes.io/arch:arm64]
aiops-quote-service-f7b5994cf-mgch6             Running   map[kubernetes.io/arch:arm64]
aiops-recommendation-service-5d57b6b77b-fhhgf   Running   map[kubernetes.io/arch:amd64]
aiops-redis-7447c4798d-bpbsq                    Running   <none>
aiops-shipping-service-698fdd4665-ljhwc         Running   map[kubernetes.io/arch:arm64]
```
The chart is deployed having some services with  `kubernetes.io/arch: arm64` as a label for the nodeSelector, some with  `kubernetes.io/arch: amd64` and some with nothing, they are all running fine


@dmitryax I agree, I will do another PR adding affinity and tolerations.

**Comment by TylerHelmuth** [comment]
@ugosan doing all three in this PR is ok since all 3 fields are related.  Maybe nest them in an object called `schedulingRules`?

**Comment by ugosan** [comment]
@TylerHelmuth all done, we have `schedulingRules.affinity` like so:

```yml
    schedulingRules:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: kubernetes.io/arch
                operator: In
                values:
                - arm64 
```
`schedulingRules.tolerations` like so:
```yml
    schedulingRules:
      tolerations:
        - key: "kubernetes.io/arch"
          operator: "Equal"
          value: "arm64"
          effect: "NoSchedule"
```

and former `nodeSelector` was moved under schedulingRules:

```yml
    schedulingRules:
      nodeSelector:
        kubernetes.io/arch: arm64
```

The default values.yml was also updated to reflect those changes

File: charts/opentelemetry-demo/values.schema.json [hunk 2]
```diff
@@ -125,6 +125,19 @@
         },
         "podAnnotations": {
           "type": "object"
+        },
+        "schedulingRules": {
+          "properties": {
```
Comment (TylerHelmuth) [inline charts/opentelemetry-demo/values.schema.json:130]:
```suggestion
        "schedulingRules": {
          "additionalProperties": false,
          "properties": {
```

File: charts/opentelemetry-demo/values.yaml [hunk 3]
```diff
@@ -72,6 +72,10 @@ components:
     podAnnotations: {}
     #  sidecar.opentelemetry.io/inject: "false"
     #  instrumentation.opentelemetry.io/inject-java: "true"
+    schedulingRules:
```
Comment (TylerHelmuth) [inline charts/opentelemetry-demo/values.yaml:75]: We can safely add `schedulingRules` to the redis component as well.  If we make it required in the values.schema.json we can also get rid of the `{{- if .schedulingRules }}` check in `_objects.tpl`.

**Review by TylerHelmuth — COMMENTED** [review]
@ugosan also please run `make generate-examples`

File: charts/opentelemetry-demo/values.yaml (hunk 3 above)
Comment (ugosan) [reply charts/opentelemetry-demo/values.yaml:75]: should we make it required? I was not so sure, I would prefer to leave the values.yml in a cleaner state if users dont have specific requirements for allocation

Comment (TylerHelmuth) [reply charts/opentelemetry-demo/values.yaml:79]: By defining it in the default values.yaml we've essentially made it required. Users would have to set it to null in their own values.yaml to get in a space where `.schedulingRules.nodeSelector` blows up on a nil `schedulingRules`.  Explicitly setting it to required avoids that situation.  I don't have strong feelings about it tho.  @dmitryax @Allex1 
 what do you think?

Comment (ugosan) [reply charts/opentelemetry-demo/values.yaml:79]: I went ahead and made it required, I was under the impression `values.yml` would need to have `schedulingRules` at least as an empty object, but no.

File: charts/opentelemetry-demo/values.yaml [hunk 4]
```diff
@@ -57,6 +57,10 @@ components:
     ports:
       - name: redis
         value: 6379
+    schedulingRules:
+      nodeSelector: {}
+      affinity: {}
+      tolerations: []
```
Comment (dmitryax) [inline charts/opentelemetry-demo/values.yaml:63]: What about putting it into default section? I believe the most common use case would be to apply the same rules to all the demo components. We can keep `schedulingRules` per component as well which would take precedence over default.

Comment (TylerHelmuth) [reply charts/opentelemetry-demo/values.yaml:63]: Great idea

Comment (ugosan) [reply charts/opentelemetry-demo/values.yaml:63]: agreed

File: charts/opentelemetry-demo/values.yaml [hunk 5]
```diff
@@ -92,11 +97,16 @@ components:
     #  sidecar.opentelemetry.io/inject: "false"
     #  instrumentation.opentelemetry.io/inject-dotnet: "true"
 
+
+
+
```
Comment (dmitryax) [inline charts/opentelemetry-demo/values.yaml]: Please avoid accidental changes like this

File: charts/opentelemetry-demo/values.yaml [hunk 6]
```diff
@@ -57,27 +61,28 @@ components:
...
+    schedulingRules: {}
     servicePort: 8080
     env:
       - name: OTEL_EXPORTER_OTLP_ENDPOINT
         value: 'http://{{ include "otel-demo.name" . }}-otelcol:4317'
       - name: AD_SERVICE_PORT
         value: "8080"
     podAnnotations: {}
-    #  sidecar.opentelemetry.io/inject: "false"
-    #  instrumentation.opentelemetry.io/inject-java: "true"
```
Comment (dmitryax) [inline charts/opentelemetry-demo/values.yaml:74]: Same here. unrelated removal

File: charts/opentelemetry-demo/values.yaml [hunk 7]
```diff
@@ -96,7 +101,9 @@ components:
     enabled: true
     useDefault:
       env: true
+
     imageOverride: {}
+    schedulingRules: {}
```
Comment (dmitryax) [inline charts/opentelemetry-demo/values.yaml:106]: Please apply consistent configs on all the components

File: charts/opentelemetry-demo/examples/default/rendered/component.yaml [hunk 8]
```diff
@@ -827,6 +834,7 @@ spec:
         app.kubernetes.io/instance: example
         app.kubernetes.io/component: frontend
     spec:
+
```
Comment (dmitryax) [inline charts/opentelemetry-demo/examples/default/rendered/component.yaml:837]: missed extra line in a template

File: charts/opentelemetry-demo/values.yaml (hunk 6 above)
Comment (ugosan) [reply charts/opentelemetry-demo/values.yaml:74]: right, sorry for the oversight

File: charts/opentelemetry-demo/values.yaml [hunk 9]
```diff
@@ -72,12 +85,17 @@ components:
     podAnnotations: {}
     #  sidecar.opentelemetry.io/inject: "false"
     #  instrumentation.opentelemetry.io/inject-java: "true"
-
+
```
Comment (TylerHelmuth) [inline charts/opentelemetry-demo/values.yaml:88]:
```suggestion

```

File: charts/opentelemetry-demo/values.schema.json [hunk 10]
```diff
@@ -330,6 +336,23 @@
...
+        },
+        "tolerations": {
+          "type": "array",
+          "items": {}
+        }
+      }
     }
   }
-}
+}
```
Comment (TylerHelmuth) [inline charts/opentelemetry-demo/values.schema.json:358]:
```suggestion
}

```

**Review by TylerHelmuth — APPROVED** [review]
Couple nits, but otherwise LGTM.  Thanks for contributing!

**Comment by TylerHelmuth** [comment]
@ugosan please rebase.

**Review by TylerHelmuth — COMMENTED** [review]
@ugosan please run `make generate-examples`.  Sorry I shouldve been more explicit when I asked you to bump the chart.  Whenever we do a chart version bump we have to regenerate the examples.

**Pull request merged**